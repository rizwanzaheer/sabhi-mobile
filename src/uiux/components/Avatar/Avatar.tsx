import * as React from 'react'
import { Image, ImageSourcePropType, PixelRatio } from 'react-native'

import Text from '../Text/Text'
import Container from '../Container/Container'
import { withTheme } from '../../theming'
import { GRAVATAR_URI } from '../../constants'
import md5 from 'md5'

interface AvatarProps {
  /**
   * Title for avatar. Gets shortened to first 2 letters. eg.'Sa' Useful for usernames.
   */
  title?: string

  /**
   * A blockchain address to create an identicon
   */
  address?: string

  /**
   * Avatar size. default 32
   */
  size?: number

  /**
   * Rounded or squared.
   */
  type?: 'circle' | 'square' | 'rounded'

  /**
   * Standard image source type prop
   */
  source?: ImageSourcePropType

  /**
   * Provide custom background color. Only visible in text avatars
   */
  backgroundColor?: string

  /**
   * Show a border around the avatar
   */
  border?: boolean

  /**
   * Specify the type of gravatar to be displayed
   */
  gravatarType?: 'identicon' | 'monsterid' | 'wavatar' | 'retro' | 'robohash' | 'blank'

  /**
   * Theme properties passed in by HOC. Can ignore.
   */
  theme: any
}

const Avatar: React.FC<AvatarProps> = props => {
  const generateTitle = (str: string) => {
    return str.charAt(0).toUpperCase() + str.charAt(1).toLowerCase()
  }

  const size = props.size ? props.size : 32
  const type = props.type ? props.type : 'circle'
  const title = props.title ? generateTitle(props.title) : '?'
  const gravatarType = props.gravatarType ? props.gravatarType : 'identicon'
  const uri =
    props.address &&
    GRAVATAR_URI +
      md5(props.address) +
      '?s=' +
      PixelRatio.getPixelSizeForLayoutSize(size) +
      '&d=' +
      gravatarType
  const avatarTypeStyle = {
    height: size,
    width: size,
    backgroundColor: props.backgroundColor ? props.backgroundColor : props.theme.colors.primary.brand,
    ...(type === 'square' ? { borderRadius: 0 } : {}),
    ...(type === 'rounded' ? { borderRadius: 5 } : {}),
    ...(type === 'circle' ? { borderRadius: size / 2 } : {}),
    ...(props.border ? { borderWidth: 2, borderColor: props.theme.colors.primary.background } : {}),
  }

  const identicon = <Image source={{ uri }} style={[avatarTypeStyle, { backgroundColor: 'transparent' }]} />
  const imageAvatar = props.source ? (
    <Image source={props.source} style={[avatarTypeStyle, { backgroundColor: 'transparent' }]} />
  ) : null
  const textAvatar = props.title ? (
    <Container viewStyle={avatarTypeStyle} alignItems={'center'} justifyContent={'center'}>
      <Text bold={true} textColor={'#ffffff'} textStyle={{ fontSize: size * 0.5 }}>
        {title}
      </Text>
    </Container>
  ) : null

  return props.source ? imageAvatar : props.title ? textAvatar : identicon
}

export default withTheme(Avatar)
