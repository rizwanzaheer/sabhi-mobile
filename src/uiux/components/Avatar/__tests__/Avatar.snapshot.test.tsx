import * as React from 'react'
import { render } from 'react-native-testing-library'
import Avatar from '../Avatar'

describe('Component(snapshots): Avatar', () => {
  it('should render with no props', () => {
    const tree = render(<Avatar />)

    expect(tree).toMatchSnapshot()
  })

  it('should render with source props', () => {
    const tree = render(<Avatar source={{ uri: 'somehttp-image-uri' }} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render with props', () => {
    const tree = render(
      <Avatar source={{ uri: 'somehttp-image-uri' }} size={32} type={'square'} border={true} />,
    )

    expect(tree).toMatchSnapshot()
  })

  it('should render with title and background props', () => {
    const tree = render(<Avatar size={32} type={'square'} title={'Satoshi'} backgroundColor={'#000000'} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render a default gravatar icon', () => {
    const tree = render(<Avatar size={32} address={'0x0f10E18715EA6F26Cf90Fd4684FdFC8c0658F9eb'} />)

    expect(tree).toMatchSnapshot()
  })

  it('should render a custom rounded gravatar', () => {
    const tree = render(
      <Avatar
        size={32}
        address={'0x0f10E18715EA6F26Cf90Fd4684FdFC8c0658F9eb'}
        gravatarType={'monsterid'}
        type={'rounded'}
      />,
    )

    expect(tree).toMatchSnapshot()
  })
})
