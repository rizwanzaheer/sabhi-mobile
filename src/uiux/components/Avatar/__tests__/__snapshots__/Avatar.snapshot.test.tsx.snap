// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshots): Avatar should render a custom rounded gravatar 1`] = `
<Image
  source={
    Object {
      "uri": "https://www.gravatar.com/avatar/77b6e2825ea079e26823f20cabbade94?s=64&d=monsterid",
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "#0078ff",
        "borderRadius": 5,
        "height": 32,
        "width": 32,
      },
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
/>
`;

exports[`Component(snapshots): Avatar should render a default gravatar icon 1`] = `
<Image
  source={
    Object {
      "uri": "https://www.gravatar.com/avatar/77b6e2825ea079e26823f20cabbade94?s=64&d=identicon",
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "#0078ff",
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      },
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
/>
`;

exports[`Component(snapshots): Avatar should render with no props 1`] = `
<Image
  source={
    Object {
      "uri": undefined,
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "#0078ff",
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      },
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
/>
`;

exports[`Component(snapshots): Avatar should render with props 1`] = `
<Image
  source={
    Object {
      "uri": "somehttp-image-uri",
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "#0078ff",
        "borderColor": "#FFFFFF",
        "borderRadius": 0,
        "borderWidth": 2,
        "height": 32,
        "width": 32,
      },
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
/>
`;

exports[`Component(snapshots): Avatar should render with source props 1`] = `
<Image
  source={
    Object {
      "uri": "somehttp-image-uri",
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "#0078ff",
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      },
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
/>
`;

exports[`Component(snapshots): Avatar should render with title and background props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000000",
      "borderRadius": 0,
      "flex": undefined,
      "flexDirection": undefined,
      "height": 32,
      "justifyContent": "center",
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": 32,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "flexWrap": "wrap",
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    Sa
  </Text>
</View>
`;
